addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,302 @@
+
+        :param str pos_order_id: The POS order to pay, as a `pos.order` id
+        :param str access_token: The access token used to verify the user
+        :param str exit_route: The URL to open to leave the POS online payment flow
+        :param dict kwargs: Data from payment module
+
+        :return: The mandatory values for the processing of the transaction
+        :rtype: dict
+        :raise: AccessError if the provided order or access token is invalid
+        :raise: ValidationError if data on the server prevents the payment
+        :raise: UserError if data provided by the user is invalid/missing
+        """
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+        self._check_is_session_open(pos_order_sudo)
+        exit_route = request.httprequest.args.get('exit_route')
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login(request)
+
+        if kwargs.get('is_validation'):
+            raise UserError(
+                _("A validation payment cannot be used for a Point of Sale online payment."))
+
+        if 'partner_id' in kwargs and kwargs['partner_id'] != partner_sudo.id:
+            raise UserError(
+                _("The provided partner_id is different than expected."))
+
+        # Don't allow passing arbitrary create values and avoid tokenization for
+        # the public user.
+        kwargs['custom_create_values'] = {
+            'pos_order_id': pos_order_sudo.id
+        }
+        if not logged_in:
+            if kwargs.get('tokenization_requested') or kwargs.get('flow') == 'token':
+                raise UserError(
+                    _("Tokenization is not available for logged off customers."))
+            kwargs['custom_create_values']['tokenize'] = False
+
+        currency_id = pos_order_sudo.currency_id
+        if not self._is_valid_currency(currency_id):
+            raise ValidationError(_("The currency is invalid."))
+        # Ignore the currency provided by the customer
+        kwargs['currency_id'] = currency_id.id
+
+        amount_to_pay = self._get_amount_to_pay(pos_order_sudo)
+        if not self._is_valid_amount(amount_to_pay, currency_id):
+            raise ValidationError(_("There is nothing to pay for this order."))
+        if tools.float_compare(kwargs['amount'], amount_to_pay, precision_rounding=currency_id.rounding) != 0:
+            raise ValidationError(
+                _("The amount to pay has changed. Please refresh the page."))
+
+        payment_option_id = kwargs.get('payment_option_id', None)
+        if not payment_option_id:
+            raise UserError(_("A payment option must be specified."))
+        flow = kwargs.get('flow', None)
+        if not flow or flow not in ['redirect', 'direct', 'token']:
```

Comment:
```suggestion
        if not (flow and flow in ['redirect', 'direct', 'token']):
```
I find it more readable this way. But if it's the opposite for you, no need to take this suggestion.

Reply:
I prefer my first approach but it's a matter of point of view and habits.